### ISSUE TITLE
e2e flake: Downward API volume should update labels on modification [Conformance]

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#17874]
In parallel suite:

```/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/downwardapi_volume.go:108 Timed out after 15.000s. Expected : content of file "/etc/labels": key1="value1" key2="value2" content of file "/etc/labels": key1="value1" key2="value2" content of file "/etc/labels": key1="value1" key2="value2" content of file "/etc/labels": key1="value1" key2="value2" content of file "/etc/labels": key1="value1" key2="value2" to contain substring : key3="value3" ```

@thockin @ixdy @wojtek-t @mikedanese @erictune

**Comment by pmorie** [comment]
@sdminonne Can you recreate this?  It could be a race or a buffer not flushing.

**Comment by sdminonne** [comment]
@pmorie @gmarek I'm having a look

**Comment by wojtek-t** [comment]
Anoter example of the failure:
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/18201/#

@sdminonne - awesome thanks!

**Comment by sdminonne** [comment]
@wojtek-t unluckily I cannot seet this page. Thanks anyway

**Comment by lavalamp** [comment]
@dchen1107 to delegate? (If you don't think it's a node bug, maybe @pmorie can find an owner?)

**Comment by karlkfi** [comment]
@sdminonne Here's our latest conformance log from master with the labels/annotations both failing (mesos/docker cluster): https://teamcity.mesosphere.io/viewLog.html?buildId=84843&buildTypeId=Oss_KubernetesMesos_ConformanceTests&tab=buildLog&guest=1

Downward API volume should update labels on modification [Conformance]
https://teamcity.mesosphere.io/viewLog.html?buildId=84843&buildTypeId=Oss_KubernetesMesos_ConformanceTests&tab=buildLog&guest=1#_focus=7087

Downward API volume should update annotations on modification [Conformance] 
https://teamcity.mesosphere.io/viewLog.html?buildId=84843&buildTypeId=Oss_KubernetesMesos_ConformanceTests&tab=buildLog&guest=1#_focus=6385

**Comment by karlkfi** [comment]
These failures can be reproduced on master with a single command:

```
./contrib/mesos/ci/test-e2e.sh -ginkgo.focus=should.update.*on.modification
```

**Comment by sdminonne** [comment]
@karlkfi thanks for the logs.  

I'm investigating the issue using `./hack/local-up-cluster.sh` (never worked with `./contrib/mesos/...` so I'm not sure if it's the same or not) 

Running only one instance of like this `
 _output/local/go/bin/e2e.test --host="127.0.0.1:8080" --provider="local" --ginkgo.v=true --ginkgo.focus="should.update.*on.modification" --kubeconfig="$HOME/.kube/config"` basically your command (I guess) I never had a failure.

Running the same command in a loop it's different and I managed to reproduce the issue, most of the time the timeout looks to aggressive 15s (while in most of the tests the timeout is something like 5 minutes).

@gmarek @wojtek-t since it's seems that this is blocking for several person I'd like to raise the timout (at least to 30seconds). I'm not saying that this is the fix (the old joke about `sleep`) but at least it should reduce the problem on busy system.

**Comment by sdminonne** [comment]
It looks to me that everytime the fails happens I ran into something like:

```
[22:58:16][Step 7/7] Dec  2 22:58:16.811: INFO: ERROR kubelet_docker_errors{operation_type="inspect_container"} => 1 @[0]
[22:58:16][Step 7/7] Dec  2 22:58:16.811: INFO: ERROR kubelet_docker_errors{operation_type="inspect_image"} => 14 @[0]
[22:58:16][Step 7/7] Dec  2 22:58:16.811: INFO: ERROR kubelet_docker_errors{operation_type="stop_container"} => 27 @[0]
```

the same is valid for @karlkfi logs.
I tried to `grep` in the code (maybe I'm blind) but I couldn't find...
Any ideas?

**Comment by wojtek-t** [comment]
@sdminonne - those lines are coming from this metric:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/metrics/metrics.go#L100

the "kubelet_" at the beginning is "subsystem" in prometheus metrics, that's why you probably didn'g find it

**Comment by gmarek** [comment]
Actually its printed here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubelet_stats.go#L100

**Comment by sdminonne** [comment]
@gmarek #18143 thanks

**Comment by wojtek-t** [comment]
We should verify if increasing the timeout:
https://github.com/kubernetes/kubernetes/pull/18143
fixed this issue.

**Comment by karlkfi** [comment]
Still fails on the mesos/docker cluster on master: https://teamcity.mesosphere.io/viewLog.html?buildId=85617&tab=buildResultsDiv&buildTypeId=Oss_KubernetesMesos_ConformanceTests&guest=1

**Comment by karlkfi** [comment]
I have been informed that the mesos executor wasn't actually handling annotation/label updates. So these tests were failing legitimately on the mesos/docker cluster. This will hopefully be fixed by #17917

**Comment by sdminonne** [comment]
@karlkfi don't know if makes sense for you but kubelet broke this and we add to fix https://github.com/kubernetes/kubernetes/pull/14557

**Comment by wojtek-t** [comment]
Another failure:
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/19139/#

**Comment by sdminonne** [comment]
:( Sorry I didn't see this... 
Ok. Having a look again (but today I'm pretty busy)

**Comment by sdminonne** [comment]
@wojtek-t https://github.com/kubernetes/kubernetes/pull/18143 is not enough. We need also to increase timeout in image which generates the logs
@gmarek sorry for this  I didn't realize it when we talked about this 
@gmarek thanks for the file!

**Comment by sdminonne** [comment]
submitted https://github.com/kubernetes/kubernetes/pull/18296 to increase timout in running images

**Comment by wojtek-t** [comment]
@sdminonne - great - thanks for looking into it!

**Comment by wojtek-t** [comment]
I'm reopening it, so that we can monitor if this really fixed the problem.
Will close it if we don't observe more errors for a couple of days.

**Comment by wojtek-t** [comment]
There weren't any failures of this test within last 3 days - closing!

Thanks @sdminonne

**Comment by pmorie** [comment]
Saw this flake again in a GKE run: https://console.developers.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/588/

**Comment by sdminonne** [comment]
@pmorie, same test but another issue.
[here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/downwardapi_volume.go#L87) we got this:
`pods "labelsupdate5f2b2573-cdc9-11e5-8c48-42010af01555" cannot be updated: the object has been modified; please apply your changes to the latest version and try again`.
While the previous issue was a too short timeout (container exited too early for slow envs).

@pmorie @wojtek-t @gmarek would you like to open another issue for this?

**Comment by sdminonne** [comment]
Working on this to propose a fix

**Comment by sdminonne** [comment]
@pmorie submitted https://github.com/kubernetes/kubernetes/pull/20815 please have a look

**Comment by sdminonne** [comment]
@wojtek-t I saw you xref this issue from #20892. I cannot see the relations (and I could be wrong).
Mind explain?
Let me know if I need to close #20815

Thanks

**Comment by wojtek-t** [comment]
@sdminonne - the test you are trying to fix failed in the presubmit of that PR:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/20892/kubernetes-pull-build-test-e2e-gce/27613/build-log.txt

I didn't review #20815 but it's not merged yet, so we don't know if that helps (hopefully yes :))

**Comment by sdminonne** [comment]
@wojtek-t I think #20815 will help on these kind of issues (retrying on the update). thanks for the explanation.

**Comment by caesarxuchao** [comment]
This test fails here for a different reason, I can create a separate issue if you think it's unrelated:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/downwardapi_volume.go:130
Expected error:
    <*errors.StatusError | 0xc208510780>: {
        ErrStatus: {
            TypeMeta: {Kind: "Status", APIVersion: "v1"},
            ListMeta: {SelfLink: "", ResourceVersion: ""},
            Status: "Failure",
            Message: "pods \"annotationupdate72f9a14b-ced8-11e5-a256-42010af0000d\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again",
            Reason: "Conflict",
            Details: {
                Name: "annotationupdate72f9a14b-ced8-11e5-a256-42010af0000d",
                Group: "",
                Kind: "pods",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
        },
    }
    pods "annotationupdate72f9a14b-ced8-11e5-a256-42010af0000d" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
not to have occurred
```
link: https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/20864/kubernetes-pull-build-test-e2e-gce/27568/

**Comment by sdminonne** [comment]
@caesarxuchao 
 In the current implementation `Update` is performed without retry. #20815 implement a retry mechanism similar to https://github.com/kubernetes/kubernetes/blob/master/test/e2e/pods.go#L440

**Comment by caesarxuchao** [comment]
@sdminonne thanks. FYI, Actually we can set the Pod.ResourceVersion="", that will force an update, it would be cleaner than using a Poll.

**Comment by sdminonne** [comment]
@caesarxuchao I'm going to update the PR with what you're suggesting. Thanks
